File: c++/leetcode208.cpp
//
// Created by Kesu Peterill on 6/17/2024.
//
#include "all.h"

class Trie {
    struct trie {
        unordered_map<char, trie> next;
        trie() {
            next.clear();
        }
    };
public:
    set<string> bottle;
    trie root;
    Trie() {

    }
    void insert(string word) {
        if (bottle.contains(word)) {
            return;
        }
        trie node = root;
        for (char c : word) {
          if (!node.next.contains(c)) {
              root.next.insert({c, trie()});
          }
          node = root.next.at(c);
        }
        bottle.insert(word);
    }

    bool search(string word) {
        return bottle.contains(word);
    }

    bool startsWith(string prefix) {
        trie node = root;
        for (char c : prefix) {
            if (!node.next.contains(c)) {
                return false;
            }
            node = root.next.at(c);
        }
        return true;
    }
};

File: java/src/leetcode208.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class leetcode208 {
    class node {
        Map<Character, node> next = new HashMap<>();
    }
    node root;
    Set<String> bottle;
    leetcode208() {
        root = new node();
        bottle = new HashSet<>();
    }
    public void insert(String word) {
        if (bottle.contains(word)) {
            return;
        }
        node n = root;
        for (int i = 0; i < word.length(); i++) {
            if (!n.next.containsKey(word.charAt(i))) {
                n.next.put(word.charAt(i), new node());
            }
            n = n.next.get(word.charAt(i));
        }
        bottle.add(word);
    }
    public boolean search(String word) {
        return bottle.contains(word);
    }

    public boolean startsWith(String prefix) {
        node n = root;
        for (int i = 0; i < prefix.length(); i++) {
            if (!n.next.containsKey(prefix.charAt(i))) {
                return false;
            }
            n = n.next.get(prefix.charAt(i));
        }
        return true;
    }
}
